geopandas/tools/sjoin.py

```diff
@@ -329,3 +333,161 @@ def _frame_join(indices, left_df, right_df, how, lsuffix, rsuffix):
             joined.index.name = right_index_name
 
     return joined
+
+
+def _nearest_query(
+    left_df: GeoDataFrame,
+    right_df: GeoDataFrame,
+    max_distance: float,
+    how: str,
+    return_distance: bool,
+):
+    if how == "inner":
+        # use whichever input is largest for the index, and the smallest for querying
+        use_left_as_sindex = len(left_df) > len(right_df)
+    else:
+        # otherwise, use the opposite of the join direction for the index
+        use_left_as_sindex = how == "right"
+    if use_left_as_sindex:
+        sindex = left_df.sindex
+        query = right_df.geometry
+    else:
+        sindex = right_df.sindex
+        query = left_df.geometry
+    if not hasattr(sindex, "nearest_all"):
```

Comment:
Since the rtree backend does not implement `nearest_all` and it has a different internal signature, it seems better to check for `PYGEOS_GE_010` as per #2053 here instead, and then raise an `UnimplementedError`.

Reply:
> rtree backend does not implement nearest_all
> and it has a different internal signature

I think these two statements are contradictory? Either rtree does not implement `nearest_all` (which is my understanding of the situation) or it does but the internal signature is different.

The goal here was to duck-type instead of essentially inspecting the backend, which I think, generally is the better approach (but that may not apply here, I'm probably missing the implications).

In any case, I think it will be very confusing for users if different spatial index backends have the same methods but have different internal signatures. I would _try_ to keep implementations cohesive (even if one of them is much slower, like we did for `query_bulk`) or name them differently somehow or not support one of them.

Reply:
Sorry for the confusion: I meant that the internal signature for `sindex.nearest` (which calls directly into `rtree.index.Index.nearest`) is [historically](https://github.com/geopandas/geopandas/pull/2053#issuecomment-895124978) different than the signature for `pygeos.strtree.STRTree.nearest`.  There are concerns in #2053 with breaking existing undocumented functionality, so the short term plan is to mark prior usage as deprecated for one release cycle, then standardize rtree nearest around the pygeos-based signature.  Completely agreed on the potential for confusion.

`nearest_all` doesn't exist for rtree but could be developed; however, if the [plan](https://github.com/geopandas/geopandas/pull/2053#issuecomment-898410907) is to drop rtree support with Shapely 2.0 (i.e., 100% pygeos-based), then I'm not sure it is worth developing it on top of rtree just for functional parity in the short term.

However, `sjoin_nearest` is new work and can work with the different backends differently to expose the same API, if appropriate.  Duck typing or a specific flag (`PYGEOS_GE_010`) have the same result; one is just a bit more explicit about why the functionality is not supported.  Which approach is better in the long term depends a lot on whether or not geopandas should have multiple, optionally-installable spatial index backends (we discussed briefly a while back, but I think the goal there was to introduce other backends like `S2`; I don't recall the specifics around `rtree`).

At the moment I think it is OK to explicitly require pygeos >= 10 in order to use a new function `sjoin_nearest` and not add similar for rtree.